pkg/util/tracing/tracer.go

```diff
@@ -646,14 +646,46 @@ func (t *Tracer) ExtractMetaFrom(carrier Carrier) (*SpanMeta, error) {
 	}, nil
 }
 
-// GetActiveSpanFromID retrieves any active span given its span ID.
-func (t *Tracer) GetActiveSpanFromID(spanID uint64) (*Span, bool) {
+// GetActiveRootSpanFromID retrieves any active span given its span ID.
+func (t *Tracer) GetActiveRootSpanFromID(spanID uint64) (*Span, bool) {
 	t.activeSpans.Lock()
 	span, found := t.activeSpans.m[spanID]
 	t.activeSpans.Unlock()
 	return span, found
 }
 
+// SetAllDescendantsSpanVerbosity takes a root Span and sets all its
+// descendant spans' verbosity using setAllChildrenSpanVerbosity. The method
+// returns true if setAllChildrenSpanVerbosity is called successfully and false
+// if not because the root Span was a no-op span.
+func (t *Tracer) SetAllDescendantsSpanVerbosity(rootSpan *Span, to bool) bool {
+	rootCrdbSpan := rootSpan.i.crdb
+	// If we have a no-op root span, we were not successful.
+	if rootCrdbSpan == nil {
+		return false
+	}
+	t.setAllChildrenSpanVerbosity(rootCrdbSpan, to)
+	return true
+}
+
+// SetAllChildrenSpanVerbosity takes a root crdbSpan, sets its children's
+// verbosity if they are not already set appropriately, and makes a recursive
+// call on its children if they have their own children.
+func (t *Tracer) setAllChildrenSpanVerbosity(rootCrdbSpan *crdbSpan, to bool) {
+	rootCrdbSpan.mu.Lock()
+	for _, child := range rootCrdbSpan.mu.recording.children {
+		if to && child.mu.recording.recordingType.load() == RecordingOff {
```

Comment:
Do you mean in this line? https://github.com/cockroachdb/cockroach/blob/d7748a971fc24936c49b5525304a01ed75d778a0/pkg/util/tracing/crdbspan.go#L104 

the `recType` checked here is the recording type passed into the argument of `enableRecording`, not the current state of the span, so I think it still needs to be checked.

Reply:
Doh, you're right. That said, I feel this kind of defensive check is best tucked within the implementation rather than at the caller. Here for e.g. we're forced to reach into `.mu.recording.recordingType` instead of blindly invoking `enableRecording`. Also, if we're conservative with our locking (as suggested below), we won't have to make these optimizations. In any case, we shouldn't make these optimizations unless we can demonstrate the need to do so through a benchmark.

Reply:
Ok, that makes sense. Thanks for explaining the logic. I'll add the defensive check to `enableRecording` and `disableRecording` for now unless you feel strongly that we should leave it out!

Reply:
(I don't feel strongly, but I suspect if you actually benchmarked here, the defensive check would not actually help much given that all it's doing down the line is grabbing a mutex + setting an atomic value - which is in the order of a few nanoseconds. Toggling the verbosity is a pretty rare operation anyway.)